default.nix

```diff
@@ -117,44 +124,53 @@ with stdenv.lib; let
 in rec {
   mkNodeModules = { src, extraEnvVars ? {} }:
     let
+      localDeps = map dirOfLocal (builtins.filter isLocal (builtins.attrValues (lock.dependencies)));
       # filter out everything except package.json and package-lock.json if possible
       # allows to avoid rebuilding node_modules if these two files didn't change
       filteredSrc =
         let
           origSrc = if src ? _isLibCleanSourceWith then src.origSrc else src;
           getRelativePath = path: removePrefix (toString origSrc + "/") path;
-          usedPaths = [ "package.json" "package-lock.json" ];
+          usedPaths = [ "package.json" "package-lock.json" ]
+            ++ localDeps;
+          dirOfRec = x: if dirOf x == "." || dirOf x == "/" then [x] else ([x] ++ dirOfRec (dirOf x));
+          usedPathsRec = builtins.concatMap dirOfRec usedPaths;
           cleanedSource = cleanSourceWith {
             src = src;
-            filter = path: type: elem (getRelativePath path) usedPaths;
+            filter = path: type: any (x: elem x usedPathsRec) (dirOfRec (getRelativePath path));
           };
         in if canCleanSource src then cleanedSource else src;
 
-      packageJson = filteredSrc + "/package.json";
-      packageLockJson = filteredSrc + "/package-lock.json";
+      peerDependencies = map (localDep: mkNodeModules { inherit extraEnvVars; src = src + ("/" + localDep); }) localDeps;
+
+      packageJson = src + "/package.json";
+      packageLockJson = src + "/package-lock.json";
       info = fromJSON (readFile packageJson);
       lock = fromJSON (readFile packageLockJson);
     in stdenv.mkDerivation ({
-      name = "${info.name}-${info.version}-node-modules";
+      name = "${normalizeName info.name}-node-modules-${info.version}";
 
       buildInputs = [ _nodejs jq ];
 
+      src = filteredSrc;
+
       npmFlags = npmFlagsNpm;
-      buildCommand = ''
+
+      phases = [ "unpackPhase" "buildPhase" ];
+      buildPhase = ''
         # do not run the toplevel lifecycle scripts, we only do dependencies
         jq '.scripts={}' ${packageJson} > ./package.json
-        cp ${packageLockJson} ./package-lock.json
-
         echo 'building npm cache'
         chmod u+w ./package-lock.json
         NODE_PATH=${npmModules} node ${./mknpmcache.js} ${cacheInput "npm-cache-input.json" lock}
 
         echo 'building node_modules'
         npm $npmFlags ci
         patchShebangs ./node_modules/
+        ${concatMapStringsSep "\n" (dep: "cp -r ${dep}/node_modules/* node_modules") peerDependencies}
```

Comment:
I'm not sure I understand the logic for doing this. AFAIU this adds dependencies of local packages to `node_modules`, but I think that we should respect the behavior of `npm install`, and if it does not include these dependencies, we should not include them either

Reply:
But if we don't include those dependencies, the resulting package won't work.

This is a good question: should we mimic the behaviour of `npm` even if it means more work needs to be done manually, or do we go for "least effort" approach (as I have done here)?

Reply:
Does that mean the package work if it's built with pure `npm`? I think we don't have to support a case like this then

Reply:
Yes, it will require additional commands if built with pure `npm` (`npm install` complains about peer dependencies and tells you how to install them manually). I don't know what's the reasoning for this. If there is a compelling reason not to automatically resolve peer dependencies this way, I'd be happy to remove this logic. However, if this npm behaviour is for historic reasons only, I think we should keep it the way it currently is becuase this is simply more user-friendly and simplifies the nix expressions needed to build the package.

Reply:
That’s not for historic reasons, it’s actually the opposite, npm < 3 used to install peer dependencies automatically, now it doesn’t. See npm/npm#6565.

What I’m somewhat confused about is that _peer_ dependencies seem to be some special weird thing meant for plugins or something like this (I have no idea). These are not the same as local (`file:`) dependencies, are they?

Reply:
Peer dependencies are transitive dependencies of local dependencies IIUC.

Reply:
Reference on peer dependencies that gets me confused: https://nodejs.org/en/blog/npm/peer-dependencies/

Reply:
Oh, ok, so I need to RTFM before writing code as it turns out.

So the peer dependency change doesn't have to do anything with the main premise of this PR -- handling `file:` dependencies. It just so happens that one of the `local` dependencies of LIGO's website has peer dependencies that aren't a dependency of the website itself.

Whether my logic of automatically installing peer deps is correct or not is a complex question, because it appears that the issues you've mentioned is to do with `npm link` and other development-time commands that don't interest us much because `nix-npm-buildpackage` is targeting CIs.

Reply:
So, IIUC, in this specific case the best solution would be to revert the peer dependency logic you added, and then... um... fix the build by either fixing the dependency that needs a peer for some reason, or adding the peer dependency to `package.yaml` of the repo?

Reply:
(I’m really with Zhenya on not doing what npm doesn’t do. I don’t really understand peer dependencies, but it kinda sounds like one’s project should declare peer dependencies of its dependencies explicitly?)